Project Side Navigation: Use EuiCollapsibleNavBeta component

tsullivan opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/162507
Relates to https://github.com/elastic/kibana/issues/166545
^ additional IA-related tasks - related to the alignment discussions - can be found here

## Work for next steps
In this PR, some work items are being saved for a next PR:
1. _Only affects Search solution_: Navigation "group titles" do not create a breadcrumb item, as sub-items in the group are not hierarchically under the title. To address this, group titles may be going away from the design. https://github.com/elastic/kibana/issues/167323
2.  _Only affects Observability solution_: Navigation accordions can not be collapsed and do not show arrow icons. To address this, in a later PR we will add internal state management for the open/closed state of each accordion. https://github.com/elastic/kibana/issues/167328
3. _Affects all solutions:_ The "collapsed" state of the side nav should show a docked view with icons-only. To address this, in later PRs we will bring Security solution into the unified nav components.
4. https://github.com/elastic/kibana/issues/167326
5. https://github.com/elastic/kibana/issues/167330
6. https://github.com/elastic/kibana/issues/167332

### Recordings
These videos show a before-and-after with the new UI. 
| project | old | new |
|--|--|--|
|observability| https://github.com/elastic/kibana/assets/908371/663765a3-4e4b-416e-b7d5-7d87eece83e8 | <img width="298" alt="CleanShot 2023-09-22 at 14 20 48@2x" src="https://github.com/elastic/kibana/assets/446285/d61f6fe0-a6a9-4806-bc27-08b0ff2afb49"> |
|search| https://github.com/elastic/kibana/assets/908371/f383773e-27a8-4485-8289-274d8231b960 | <img width="281" alt="CleanShot 2023-09-22 at 14 18 43@2x" src="https://github.com/elastic/kibana/assets/446285/901b8fc1-9945-4cee-b566-5e4539f08043"> |
|security| https://github.com/elastic/kibana/assets/908371/481f4533-64e5-41db-bc8e-5012f82c188a | *will change to the new style after this PR and the flyout/panel support are completed |

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

tsullivan committed:
fix unused prop isOpen for app menu bar

tsullivan committed:
remove custom css additions

tsullivan committed:
Use EuiCollapsibleNavBeta stories

tsullivan committed:
Fix recently accessed bucket

tsullivan committed:
replace EuiCollapsibleNavGroup with EuiCollapsibleNavItem & EuiSideNavItemType with EuiCollapsibleNavSubItemProps

tsullivan committed:
Replace EuiFlexGroups with EuiFlyoutBody & EuiFlyoutFooter

tsullivan committed:
Remove empty root items from storybook

tsullivan committed:
Remove empty title parents from preset links

tsullivan committed:
Add silly ts-expect-error

tsullivan committed:
Remove empty groups from stories

tsullivan committed:
Incorporate isGroupTitle

tsullivan committed:
remove close button wrappers in core chrome-browser

tsullivan committed:
fix hierarchy of svl search project settings

tsullivan committed:
fix hierarchy of observablility side nav

tsullivan commented on packages/shared-ux/chrome/navigation/src/ui/components/navigation_section_ui.tsx:
```diff
@@ -39,6 +37,7 @@ const navigationNodeToEuiItem = (
     [`nav-item-isActive`]: isSelected,
   });
 
+  // @ts-expect-error getRenderItem is not used (doesn't exist on EuiCollapsibleNavSubItemProps)
```
FIXME: Not sure what this was originally doing, but it probably shouldn't just be unused

cee-chen replied:
my (contextless) 2c: I'd suggest removing it entirely, since the new component won't use it.

tsullivan replied:
@cee-chen 

Seb caught the problem with this removal here: https://github.com/elastic/kibana/pull/164910#discussion_r1314588325

This was providing functionality that allowed teams to render any content inside the navigation node. It's important we're still able to do this: the Security Solution team plans to start using our navigation node types, and include the flyout menus that we're familiar with. To do that, I believe we need [the `renderItem` support that exists in EuiSideNavItemType](https://github.com/elastic/eui/blob/d4561360a99324ce1bc3c65b7f6c00ebdae17df1/src/components/side_nav/side_nav_types.ts#L39-L42)

@sebelga can you confirm this?

cee-chen replied:
Haha, we're mindmelding right now. I responded at the exact same time with a comment down below: https://github.com/elastic/kibana/pull/164910#discussion_r1326632602

I think the only thing that's not totally clear to me is whether Kibana needs the ability to custom render just top level items (if so: doesn't need an EUI enhancement), or if it needs the ability to custom render nested items (if so: does need an EUI enhancement).

tsullivan replied:
In a separate PR, we will update the API for declaring the navigation tree to allow consumers to declare flyouts. The props for the API change will be handled at a higher level: if a navigation item declares a flyout, a wrapper around our `EuiCollapsibleNavItem` will defer to another way of rendering the secondary action and enabling the flyout. 

Added 1f95d3c9d9d

cc @sebelga

tsullivan commented:
ping @elastic/kibana-design

cee-chen commented on packages/core/chrome/core-chrome-browser-internal/src/ui/project/header.tsx:
```diff
@@ -179,9 +167,6 @@ export const ProjectHeader = ({
   ...observables
 }: Props) => {
-  const [navId] = useState(htmlIdGenerator()());
-  const [isOpen, setIsOpen] = useLocalStorage(LOCAL_STORAGE_IS_OPEN_KEY, true);
```
FYI - you can still store/fetch this information in localStorage using `<EuiCollapsibleNavBeta initialIsCollapsed={someLocalStorageData}>`. Just want to make sure we're not losing any UX in the new nav.

... Although on reflection, I never added a `onToggleCollapse` callback to the new component 🤦 Let me add that real quick...

tsullivan replied:
Thanks so much @cee-chen! I was able to use your latest code in 9ddf71b4e88

cee-chen replied:
Perfection!!!

MichaelMarcialis commented:
Thanks so much for these comparison videos, @tsullivan. They were very helpful. And thanks again to you and @cee-chen for pulling this all together so quickly. Generally speaking, this is looks wonderful. I'm leaving a few quick thoughts/comments below based on my initial viewing of these videos.

#### General

- For the sake of the applications that are very UI-dense and require as much horizontal real estate as possible (ex. Discover, Lens, etc.), our original suggestion was that the default state of the navigation should be collapsed. If/when users wish to have a larger navigation present, they could then switch to the expanded mode. Currently, it appears that the expanded navigation is the default. Would it be possible to make the collapsed state the default?

- Our original intent with the new navigation was that only a single top-level accordion should be opened at a time. The thinking was that doing so would 1) avoid potentially overwhelming users with a large number of links and 2) avoid presenting a navigation that has multiple vertical scrollbars in play at once. Would it be possible to make it so that opening a top-level accordion closes any other open top-level accordions automatically?

- It appears that the individual navigation links are currently limited to a single line of text. Exceeding this triggers ellipsis truncation. Ideally we'd allow navigation text to break a line if needed to avoid such truncation. Would that be possible to change, or is there a compelling technical reason to restrict the links to a single line?

- This is tangential to the side navigation, but I wanted to mention it to make sure it wasn't lost/forgotten. Do we still plan to update the breadcrumbs to 1) use the cleaner looking `type="page"` prop and 2) incorporate the project and space (assuming space is still a thing for serverless) buttons as the first two items of the breadcrumbs?

#### Observability Solution

- Two links under the "AIOps" accordion appear to be using a prepended `beaker` icon. I assume these are meant to identify these links as features under "Technical preview". If that assumption is correct, I have a few minor concerns. 
    1. The original intent for prepended navigation icons was that they would only be used to help identify the various saved object types under the "Recent" section. If it's not absolutely necessary to inform the users of feature status within the navigation, could we instead inform users of the status on the actual pages? If it is absolutely necessary, perhaps an appended badge component would be a better choice (assuming the navigation supports such an append)?
    2. Are these features actually labelled "Technical preview"? Or are they "Experimental"? Asking because if it's "Technical preview" the `beaker` icon no longer makes much sense. A beaker (or flask in reality) doesn't convey that wording to users. 

#### Search Solution

- I noticed that "Developer tools" is not a top-level navigation item as the designs had suggested. I imagine this was an intentional decision made by the Search solution product team, but I wanted to call out the inconsistency when compared to the other solutions here for the sake of visibility.

- It looks like there's only one link under "Project settings" for this solution. We were anticipating more than that (namely links to the cloud portal, such as user and roles, performance, billing and subscription). Were those additional links intentionally omitted? If intentional, would be better to change the top-level accordion to a simple link that points to the management hub page (rather than an opening an accordion with just that link)?

#### Security Solution

- It appears that Security will need to either 1) come up with a technical solution that doesn't cause the navigation to break in collapsed mode, or 2) decide to (at least temporarily) conform to the navigation we've provided and use it the way it was intended.

CCing @ryankeairns, in case he has any additional comments. Thanks!

sebelga commented:
> Would it be possible to make it so that opening a top-level accordion closes any other open top-level accordions automatically?

@MichaelMarcialis I just want to point out that by doing this the user would lose the "active" state of the page they are currently in.

> It looks like there's only one link under "Project settings" for this solution. We were anticipating more than that (namely links to the cloud portal, such as user and roles, performance, billing and subscription). Were those additional links intentionally omitted?

For those links to appear, their URLs need to be added in the `serverless.yml`. Tim probably didn't do it in his local env, but those are injected in the serverless deployment.

sebelga commented on packages/core/chrome/core-chrome-browser-internal/src/ui/project/navigation.tsx:
```diff
@@ -7,54 +7,16 @@
-        /* must render the tree to initialize the navigation, even if it shouldn't be visible */
-        !isOpen && <div hidden>{children}</div>
+        /* must render the tree to initialize the navigation, even if hidden internally in EUI */
+        <div hidden>{children}</div>
```
I think it would be better to use `display:none` or `display:block` on `EuiCollapsibleNavBeta` and remove those L16 and L17 to avoid re-rendering the nav tree every time.

tsullivan replied:
Looking back in the history of this file, this code was added to solve `Issue 3. Navigation is not initialized when Kibana loaded with hidden navigation` in https://github.com/elastic/kibana/pull/163700

tsullivan replied:
> I think it would be better to use display:none or display:block on EuiCollapsibleNavBeta and remove those L16 and L17 to avoid re-rendering the nav tree every time.

@sebelga this will work. The old code used `isOpen && children` to stop rendering when the nav is collapsed, which had a bug where the navigation tree would not get initialized when you refresh the page with hidden navigation, causing breadcrumbs and maybe other things to not work. (@Dosant can you confirm I understand that issue correctly?) The new code doesn't have that problem since the navigation is hidden using CSS. 

Added 879b29fd910 to clean that up, and also also unblocks this PR regarding a separate issue.

sebelga commented on packages/shared-ux/chrome/navigation/src/ui/components/navigation_section_ui.tsx:
```diff
@@ -65,14 +66,13 @@ const navigationNodeToEuiItem = (
         : undefined,
     href,
-    renderItem: getRenderItem(),
+    // renderItem: getRenderItem(),
```
This functionality allowed teams to render any content inside the navigation node for maximum flexibility. Do we plan to keep allowing it?

tsullivan replied:
I believe we must keep it, since that plan has not changed. I think support for this must be carried into `EuiCollapsibleNavSubItemProps` as commented above.

cee-chen replied:
@tsullivan Can you clarify on this statement? Are you saying you need me to update `EuiCollapsibleNavItem` on the EUI side of things with a new `renderItem` prop?

Or alternatively, if this is meant to render custom top-level items, could you make this a conditional replacement on Kibana's side, e.g.

```tsx
{getRenderItem ? getRenderItem() : <EuiCollapsibleNav {...props} />}
````

tsullivan replied:
I need to sync with Seb on this, will get back to you next week.

sebelga commented on packages/shared-ux/chrome/navigation/src/ui/components/navigation_ui.tsx:
```diff
@@ -21,17 +21,12 @@ export const NavigationUI: FC<Props> = ({ children, unstyled, footerChildren, da
-          {footerChildren && <EuiFlexItem grow={false}>{footerChildren}</EuiFlexItem>}
-        </EuiFlexGroup>
+        <>
+          <EuiFlyoutBody scrollableTabIndex={-1} data-test-subj={dataTestSubj}>
```
I'm surprised to see a flyout component used here? Why did you decide to make this change?

cee-chen replied:
To clarify, `EuiCollapsibleNav` (both the old component and the new beta component) have always used `EuiFlyout` under the hood. So using a flyout body and footer as well feels totally natural to me. I also wanted to leave `children` as flexible as possible for usage, but if this is confusing to folks, one option is we could export the the flyout body & footer as `EuiCollapsibleNav.Body` and `EuiCollapsibleNav.Footer`

sebelga replied:
I was just wondering why we can't simply have 2 flex boxes that expand as I put it together in the current version. What does the Flyout component brings?

cee-chen replied:
To clarify, the footer **doesn't** expand while the body does (and also flyout bodies automatically adds a scrollbar if content is too tall).

BTW, if you're not using the footer behavior at all, it's totally valid to skip both. The collapsible nav will account for that.

> What does the Flyout component brings?

Some extra styling:

https://github.com/elastic/eui/blob/829709eebc261132af96a154757b6687a31719bd/src/components/collapsible_nav_beta/collapsible_nav_beta.styles.ts#L36-L44

To be clear, you could definitely recreate this yourself if you have to - the CSS isn't terribly complex. But it's just an extra set of small things that get handled out of the box for you by EUI if you do.

tsullivan replied:
I followed Cee's examples in the EUI storybook, and it seems to be the right way to go from my point-of-view as well. I'm ++ for the scrollbar behavior

sebelga replied:
Cool, thanks for the explanation 👍

loixlab reviewed (commented):
Great work @tsullivan ! Overall looks good 👍 I left a few comments in the code to follow up.

Visually I am not a big fan of CSS transition in menus. The "clipping" that occurs is not so nice. I prefer a more "snappy" experience.
Also, I'd probably add a little more vertical space between the "level 2" toggle groups.

But that's personal taste and I leave it to the design team. 😊

MichaelMarcialis commented:
> > Would it be possible to make it so that opening a top-level accordion closes any other open top-level accordions automatically?
> 
> @MichaelMarcialis I just want to point out that by doing this the user would lose the "active" state of the page they are currently in.

@sebelga: In the navigation designs, we accounted for that situation by transitioning the "current" state from the individual link to the closest visible parent item. In the case that an entire top-level accordion is closed (as suggested above) while viewing a page within that section, the "current" state should be applied to that top-level parent to inform the user what section they are in ([see example in Figma mockups](https://www.figma.com/file/8X4biZxZ4mkcDpYFCODRcV/IC2-Mockups-%2F-Version-2-%5BIn-Progress%5D?type=design&node-id=960%3A348206&mode=design&t=CXooHRRi1UgYq2Nz-1)). If that behavior is not something that's already happening in the serverless navigation, would it be possible to add it?

cee-chen commented:
> Visually I am not a big fan of CSS transition in menus. The "clipping" that occurs is not so nice. I prefer a more "snappy" experience.

I'm a little confused about what you mean by this - I'm not seeing any clipping on the EUI side of things. Do you mind including a video screencap of what you're seeing?

sebelga commented:
> the "current" state should be applied to that top-level parent to inform the user what section they are in. If that behavior is not something that's already happening in the serverless navigation, would it be possible to add it?

@MichaelMarcialis  Ah yes I didn't see the mockup. It should be possible to highlight the group, the active state contains each node of the branch that matches the URL.

> Do you mind including a video screencap of what you're seeing?

@cee-chen here is a gift of the clipping. You can see that it is not "snappy" and transitions in 2 phases

![nav-clip](https://github.com/elastic/kibana/assets/2854616/e7ddf649-cd87-4fb7-861e-8671a6fa9ebb)

cee-chen commented:
> @cee-chen here is a gift of the clipping. You can see that it is not "snappy" and transitions in 2 phases

Super interesting! Do you mind sharing what browser and OS you're on? That isn't something I've seen happen personally with accordion expansion, and I'm wondering what the cause could be for it. Michael or Tim, can you repro the gif Sébastien included above?

sebelga commented:
> Do you mind sharing what browser and OS you're on? 

Mac OS `12.6.8`, Chrome `116.0.5845.140`

MichaelMarcialis commented:
> Michael or Tim, can you repro the gif Sébastien included above?

I tried to run this branch locally, but Docker keeps saying I don't have access to pull what it needs (or it doesn't exist). Is there something I need to do to get this running locally?

ryankeairns commented:
Hey @tsullivan , as part of this PR, can you change the defaults on the solution nav group so that it is always open (i.e. has no collapse option/icon)? Thanks!

tsullivan committed:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

tsullivan committed:
expose accordionProps and disable collapsing in main solution sections

tsullivan commented:
@ryankeairns 

> can you change the defaults on the solution nav group so that it is always open (i.e. has no collapse option/icon)? Thanks!

Sure! Done in c449de208e9.

cee-chen commented on x-pack/plugins/serverless_search/public/layout/nav.tsx:
```diff
@@ -33,6 +33,10 @@ const navigationTree: NavigationTreeDefinition = {
       defaultIsCollapsed: false,
+      accordionProps: {
+        forceState: 'open',
+        arrowProps: { css: { display: 'none' } },
```
```suggestion
        arrowDisplay: 'none',
```

tsullivan replied:
@cee-chen I tried this out, and I'm seeing the arrow still visible, yet moved to the left.

Before:
<img width="286" alt="image" src="https://github.com/elastic/kibana/assets/908371/992e74bd-11d3-4d22-9692-c3d8f479bce0">


After:
<img width="313" alt="image" src="https://github.com/elastic/kibana/assets/908371/c2a0aa55-6de6-4d6c-8c92-d2ceb82f8ed6">

cee-chen replied:
🤦 That's a bug in EUI, I'll get a fix out

cee-chen replied:
Ah apologies, it's not a bug, it's simply that accordions don't allow the arrow to be hidden if the button isn't interactive. Essentially, there always needs to be a way to toggle the accordion.

@ryankeairns 

> defaults on the solution nav group so that it is always open (i.e. has no collapse option/icon)?

Can you rethink this ask? If it's truly crucial, the component used essentially **can't** be an accordion. Hiding the arrow is fine for visual users but there's still a ton of accessibility implications of accordions that's going to confuse screen reader users if we leave it in.

If we need a way to render a header with items that's not collapsible, that's fine, but I need to render it a different way than using an underlying `EuiAccordion`.

ryankeairns replied:
> If we need a way to render a header with items that's not collapsible, that's fine, but I need to render it a different way than using an underlying EuiAccordion.

10-4. I'll come back around to this with an EUI issue.

ryankeairns replied:
Created this feature request in EUI: https://github.com/elastic/eui/issues/7208

tsullivan committed:
track collapsed state in local storage

tsullivan commented:
> Michael or Tim, can you repro the gif Sébastien included above?

I noticed the clipping behavior that Seb called out, but only in the initial development phase. Now that I'm looking again 2 weeks later, I don't see it anymore. There's no transition as child items are shown, they just "snap" into view. Maybe some style changed behind the scenes?

cee-chen commented:
> As of now there is no transition as child items are shown, they just "snap" into view. Maybe some style changed behind the scenes?

Haha oh jeez, that shouldn't be happening either 🙈 It should (ideally, given a non-overloaded GPU) animate the height of the accordion, although of course anytime you animate height it's unperformant and is simultaneously causing a repaint and a reflow so for lower-power machines it can definitely appear choppy.

It might be worth scheduling a sync Zoom call for this to share our screens and try to figure out what each other is seeing.

tsullivan commented:
@cee-chen I agree to meet and discuss. I see what you are describing in the EUI repo's storybook, based on the current main branch. It looks very smooth and polished in storybook. Perhaps there is an integration we can figure out. 

@MichaelMarcialis 
> I tried to run this branch locally, but Docker keeps saying I don't have access to pull what it needs (or it doesn't exist). Is there something I need to do to get this running locally?

The steps to run ES and Kibana in serverless mode locally:
1. Type `yarn es serverless`
2. type `yarn serverless` in another terminal. Once Kibana is running, you can log in with `elastic_serverless / changeme`.

In dev mode only, there is a project switcher that lets you choose different project types.

You can have serverless-specific settings for your Kibana dev environment in a `config/serverless.dev.yml` file. In this file, you would want to add a cloud ID such as: `xpack.cloud.id: my_cool_fake_cloud_id`

tsullivan committed:
Merge remote-tracking branch 'elastic/main' into serverless-chrome/use-EuiCollapsibleNavBeta

tsullivan committed:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

tsullivan committed:
Completely hide the side bar in collapsed state

This commit takes away the "dock" which is the side bar in its
"collapsed state." This hiding is temporary until all solution
integrations are using the same navigation tree component.

tsullivan committed:
Remove renderItem support

In a separate PR, we will update the API for declaring the navigation
tree to allow consumers to declare flyouts. The props for the API change
will be handled at a higher level: if a navigation item declares a
flyout, a wrapper around our `EuiCollapsibleNavItem` will defer to
another way of rendering the secondary action and enabling the flyout.

tsullivan committed:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

tsullivan committed:
fix warnings in test file

tsullivan committed:
fix unit tests

tsullivan committed:
Add FIXME item for expanding active accordions

tsullivan committed:
fix more unit tests

tsullivan committed:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

tsullivan committed:
fix solutions group titles

tsullivan committed:
fix more tests

tsullivan committed:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

tsullivan committed:
update storybooks to give an updating state example

tsullivan committed:
wip

tsullivan committed:
More storybook

tsullivan committed:
wip2

tsullivan committed:
Updating state storybook

tsullivan committed:
cleanup

tsullivan committed:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

tsullivan commented on x-pack/plugins/serverless_observability/public/components/side_navigation/index.tsx:
```diff
@@ -25,136 +25,129 @@ const navigationTree: NavigationTreeDefinition = {
-            },
-          ],
+          id: 'groups-spacer-1',
+          isGroupTitle: true,
```
I chatted with @cee-chen about the need for a nav item prop of `isSpacer`, which would improve this syntax aesthetically, with something like: `{ isSpacer: true }`

sebelga replied:
Not sure why why we would need an extra DOM element. We should be able to add css to previous group to create spaces. The same way we add gutter between flex elements.

sebelga replied:
> do you choose to be messy in your CSS, or do you choose to be messy in your DOM/JSX

I disagree here. This is not a question of being messy or not. Designing a navigation with items groups under different section should not have an unlimited number of options to space the groups. "Small", "medium" and "large" should be enough. There is only one side nav in an app at the end of the day 😊 And we have a very specific design we are aiming to build in Serverless.

Using DOM nodes to add spaces is, IMO, not a good choice: it affect semantic (a DOM element is used purely for layout purposes), it affects performance (large DOM trees negatively affect performance).

> For more context, there isn't a concept of "groups" anymore in the new component - an item is either an accordion or a link

This is something we might want to revisit. We discussed with the team yesterday, we might still want to be able to group navigation items without necessarily have an accordion. For example to render a bold header with a few links under it.
This should be possible by adding a few prop to the "group" (e..g `collapsible: false` --> this group is not collapsible, don't render an accordion, make the title bold and align the items on the left)

sebelga replied:
To not block the work on this PR let's leave it like that. I will probably have a second pass at this. I don't think the solution teams should be adding fake "group-spacer" nodes in they navigation tree definition to create a nice balanced space between groups. The layout and spacing should be handled by our shared components.

One team might be tempted to add 2 fake nodes, another team 1, and before we know it all navigation look different...

tsullivan replied:
I think the second pass that Seb is commenting about would just add capabilities to the Shared UX side nav package in Kibana, and not require changes made in EUI, right? I think we all agree that we don't want EUI to be over-specialized.

cee-chen replied:
The `items` prop API of `EuiCollapsibleNavItem` is overly specialized currently - what I'll probably do is add a new generic `renderItem: () => {}` type/key to it (that replaces `isGroupTitle`) so that y'all can render whatever you need for groups and so forth (whether it's a title with a specific padding, or a specific spacer size with no title).

I'm also working on a new `EuiCollapsibleNavBeta.Group` component for you to use for "accordions" that are always open, that you can customize the spacing of as needed.

Also just to clarify, but none of these new APIs need to hold up this PR, I'm just trying to work in things into the base EUI components to make your lives easier!

tsullivan commented on packages/shared-ux/chrome/navigation/src/ui/components/navigation_section_ui.tsx:
```diff
@@ -39,24 +34,18 @@ const navigationNodeToEuiItem = (
+    title: item.title,
     isSelected,
+    accordionProps: {
+      initialIsOpen: true, // FIXME open state is controlled on component mount
```
In this PR, all sub-accordions in a nav tree will be expanded by default: `initialIsOpen` controls whether the accordion items of EuiCollapsibleNavItem are open by default on component mount.

More work is needed after this PR to have the Kibana component to take control of open/closed state of accordion nav items with the "active nodes" state we track in this component.

With "active nodes" state changes, accordion items the user explicitly opened need to remain open, even if none of the items are selected - that means extra state we need to track in this component.

cc @cee-chen @sebelga

sebelga replied:
Ok. I'll have a look into it in a future PR. For now keeping them open works 👍

tsullivan replied:
Filed https://github.com/elastic/kibana/issues/167328

tsullivan commented on packages/shared-ux/chrome/navigation/src/ui/components/navigation_section_ui.tsx:
```diff
@@ -147,30 +138,25 @@ export const NavigationSectionUI: FC<Props> = ({ navNode, items = [] }) => {
       {...propsForGroupAsLink}
-    >
-      <EuiText color="default">
-        <EuiSideNav
```
Note: `EuiSideNav` was previously doing some opinionated work in the background automatically to expand accordion items that had `isSelected` children. Now that we're not using `EuiSideNav`, we need to move that work to the Kibana side. See https://github.com/elastic/kibana/pull/164910/files#r1334691666

tsullivan committed:
silence a minor TS error

cc @TattdCodeMonkey

tsullivan commented on packages/shared-ux/chrome/navigation/src/ui/navigation.stories.tsx:
```diff
@@ -551,3 +514,121 @@ export const CreativeUI = (args: NavigationServices) => {
   );
 };
+
+export const UpdatingState = (args: NavigationServices) => {
```
I created this new Storybook for manually testing state updates to the active nodes. I wasn't really sure how to use Storybook actions or argTypes to do this: I updated the template component to show a button and put the click handler in the Storybook component to update state in the mocked services.

@clintandrewhall would you please take a look and let me know if you see a better way to do this?

tsullivan commented on x-pack/plugins/serverless_search/public/layout/nav.tsx:
```diff
@@ -17,6 +17,7 @@ import type { ServerlessPluginStart } from '@kbn/serverless/public';
 import type { CloudStart } from '@kbn/cloud-plugin/public';
 
 // Hiding this until page is in a better space
+// @ts-expect-error: variable is not used
```
FYI @TattdCodeMonkey

TattdCodeMonkey replied:
Thanks!

TattdCodeMonkey reviewed (approved):
Serverless Search nav changes LGTM.

cc: @julianrosado to 👀  and be aware of the new UX

tsullivan committed:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

tsullivan committed:
update navigation isSectionOpen ftr function

tsullivan committed:
temporary search nav tree fix for functional tests

tsullivan committed:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

tsullivan committed:
Revert "temporary search nav tree fix for functional tests"

This reverts commit 2ba9ba3f8c0dda5639a991faeef778776ac75002.

tsullivan committed:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

alisonelizabeth reviewed (approved):
Deployment Management changes LGTM. Code review only.

tsullivan committed:
fix serverless search navigation tests

tsullivan committed:
fix serverless search navigation tests

tsullivan committed:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

tsullivan committed:
temporarily remove ability to close accordions

tsullivan committed:
fix EuiCollapsibleNav as button

tsullivan committed:
Merge branch 'serverless-chrome/use-EuiCollapsibleNavBeta' of github.com:tsullivan/kibana into serverless-chrome/use-EuiCollapsibleNavBeta

davismcphee reviewed (approved):
Code-only review. `packages/default-nav/analytics/default_navigation.ts` changes LGTM 👍

sorenlouv reviewed (approved):
apm (aka serverless_observability) changes lgtm

loixlab reviewed (commented):
Amazing work @tsullivan ! 🎉 I tested locally and all works as expected.

I left a few comments, mostly those will be addressed as follow-up work. What I would like to fix in this PR are the tests as suggested in my comments. Let me know if it is really not possible to have the previous assertions. Cheers!

### Follow-up work

* It seems that the new `accordionProps` is only used to hide the accordion arrow with css. I will probably refactor it to not exposed those props and have a flag to mark a group as collapsible or not. Unless we find value in allowing full access to those props.
* The Initial open state wraps the area as active on page load. Can we prevent that behaviour from happening and remove the highlight on page load (basically keep the active element at the top left corner)? @cee-chen 
<img width="265" alt="Screenshot 2023-09-26 at 13 59 27" src="https://github.com/elastic/kibana/assets/2854616/3a392018-7a98-4e09-b053-6297f9f064e9">

* Could we remove the animation when toggling on/off the side bar? I would find it more "snappy" than this "long" transition.

![slow-appear-menu](https://github.com/elastic/kibana/assets/2854616/297b050e-a000-4d4d-a56b-2b1054d1e5c6)

* Where we currently have the bold title for groups of items (Search nav bar) I will try to make the "grouping" work with the breadcrumb by nesting the `isGroupTitle` node with its underlying `children`.

tsullivan committed:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

drewdaemon reviewed (approved):
LGTM, code review only

tsullivan commented on x-pack/plugins/serverless_observability/public/components/side_navigation/index.tsx:
```diff
@@ -25,136 +25,135 @@ const navigationTree: NavigationTreeDefinition = {
+          title: i18n.translate('xpack.serverlessObservability.nav.applications', {
+            defaultMessage: 'Applications',
+          }),
+          accordionProps: {
```
To clarify, ideally this section would be collapsible and show the arrow. The reason for now showing the arrow in this iteration is due to the lack of proper state management that keeps the necessary accordions expanded. To move forward without the proper state management, I am making all sections expanded and non-collapsible, and that is just to avoid problems with using nav in this intermediary state. The comments about this are here: https://github.com/elastic/kibana/pull/164910#discussion_r1337164080

tsullivan replied:
related: https://github.com/elastic/kibana/issues/167328

tsullivan committed:
remove debug comments

tsullivan committed:
update navigation tree tests with .toBeVisible checks

tsullivan committed:
update ML default navigation to have no root parent for top-level items

tsullivan committed:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

tsullivan committed:
update snapshot

tsullivan commented on packages/shared-ux/chrome/navigation/src/ui/components/navigation.test.tsx:
```diff
@@ -427,10 +413,6 @@ describe('<Navigation />', () => {
         jest.advanceTimersByTime(SET_NAVIGATION_DELAY);
       });
 
-      expect(await findByTestId('my-custom-element')).toBeVisible();
```
This test depended on navigation nodes supporting the `renderItem` prop, which isn't supported any longer in `EuiCollapsibleNavItem`. I'll go ahead and delete this test.

loixlab reviewed (approved):
Thanks for making the changes 👍

Commit:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

tsullivan commented:
@elasticmachine merge upstream

tsullivan committed:
Merge branch 'main' into serverless-chrome/use-EuiCollapsibleNavBeta

alvarezmelissa87 reviewed (approved):
ML code changes LGTM
